python_modules/dagster-external/dagster_external/context.py

```diff
@@ -0,0 +1,126 @@
+    read_mode = os.getenv(DAGSTER_EXTERNAL_ENV_KEYS["read_mode"], "stdin")
+    if read_mode == "stdin":
+        input_stream = sys.stdin
+    elif read_mode == "fifo":
+        fifo_path = os.getenv(
+            DAGSTER_EXTERNAL_ENV_KEYS["input"], DEFAULT_DAGSTER_EXTERNAL_INPUT_FIFO
+        )
+        input_stream = open(fifo_path, "r")
+    else:
+        raise Exception(f"Invalid read mode: {read_mode}")
+
+    write_mode = os.getenv(DAGSTER_EXTERNAL_ENV_KEYS["write_mode"], "stdout")
+    if write_mode == "stdout":
+        output_stream = sys.stdout
+    elif write_mode == "fifo":
+        fifo_path = os.getenv(
+            DAGSTER_EXTERNAL_ENV_KEYS["output"], DEFAULT_DAGSTER_EXTERNAL_OUTPUT_FIFO
+        )
+        output_stream = open(fifo_path, "w")
+    else:
+        raise Exception(f"Invalid write mode: {write_mode}")
+    data = json.loads(input_stream.read())
+    context = ExternalExecutionContext(data, output_stream)
+    ExternalExecutionContext.set(context)
+    return context
+
+
+class ExternalExecutionContext:
+    _instance: ClassVar[Optional[Self]] = None
+
+    @classmethod
+    def set(cls, context: Self) -> None:
+        cls._instance = context
+
+    @classmethod
+    def get(cls) -> Self:
+        if cls._instance is None:
+            raise Exception(
+                "ExternalExecutionContext has not been initialized. You must call"
+                " `init_dagster_external()`."
+            )
+        return cls._instance
+
+    def __init__(self, data: ExternalExecutionContextData, output_stream: IO) -> None:
+        self._context = None
+        self._data = data
+        self._output_stream = output_stream
+
+        self._channel: Optional[IO[str]] = None
+
+    def _send_notification(self, method: str, params: Optional[Mapping[str, Any]] = None) -> None:
+        notification = Notification(method=method, params=params)
+        self._output_stream.write(json.dumps(notification) + "\n")
+
+    # ########################
+    # ##### PUBLIC API
+    # ########################
+
+    @property
+    def asset_key(self) -> str:
+        return self._data["asset_key"]
+
+    @property
+    def code_version(self) -> Optional[str]:
+        return self._data["code_version"]
+
+    @property
+    def data_provenance(self) -> Optional["ExternalDataProvenance"]:
+        return self._data["data_provenance"]
+
+    @property
+    def partition_key(self) -> Optional[str]:
+        return self._data["partition_key"]
+
+    @property
+    def partition_key_range(self) -> Optional["ExternalPartitionKeyRange"]:
+        return self._data["partition_key_range"]
+
+    @property
+    def partition_time_window(self) -> Optional["ExternalTimeWindow"]:
+        return self._data["partition_time_window"]
+
+    @property
+    def run_id(self) -> str:
+        return self._data["run_id"]
+
+    @property
+    def tags(self) -> Mapping[str, str]:
+        return self._data["tags"]
+
```

Comment:
nit: run tags vs op tags , not sure its safe to count on this only being used in assets or assets never adding tags to propagate to their op